File: ds4-gateway/genie_power.py
"""Chat access to the enrolled fleet power scripts via the dashboard tool endpoint."""
import json
import urllib.error
import urllib.parse
import urllib.request
TOOLSET='stargate_fleet_power'
NAMES={'fleet_power_status','fleet_power','fleet_routing','fleet_recipe_trial','fleet_recipe_rollout'}


def register_power(config,emit):
    from tools.registry import registry
    url=urllib.parse.urlsplit(config['url'])
    if (url.scheme!='http' or url.hostname!='127.0.0.1' or not url.port or url.path!='/api/genie/power-tools'
            or url.username or url.password or url.query or url.fragment):
        raise ValueError('Use the private fleet power tool endpoint')
    class NoRedirect(urllib.request.HTTPRedirectHandler):
        def redirect_request(self,*args,**kwargs):return None
    opener=urllib.request.build_opener(urllib.request.ProxyHandler({}),NoRedirect())
    def run(name,args):
        event={'tool':name,'at':__import__('datetime').datetime.now(__import__('datetime').timezone.utc).isoformat(),'request':args}
        emit('power',event={**event,'state':'reading'})
        try:
            if name=='fleet_power_status':payload={'action':'status'}
            elif name=='fleet_recipe_trial':payload={'action':'recipe-trial','profile':args['profile'],'stage':args['stage'],'trial_id':args['trial_id']}
            elif name=='fleet_recipe_rollout':payload={'action':'recipe-rollout','profile':args['profile'],'rollout_id':args['rollout_id'],**({'expected_finished_at':args['expected_finished_at']} if 'expected_finished_at' in args else {})}
            elif name=='fleet_routing':payload={'action':'routing','worker':args['worker'],'routing_action':args['routing_action'],'action_id':args['action_id'],**({'expected_operator_action':args['expected_operator_action']} if 'expected_operator_action' in args else {})}
            else:payload={'action':'power','worker':args['worker'],'power_action':args['power_action'],'action_id':args['action_id']}
            request=urllib.request.Request(config['url'],data=json.dumps(payload).encode(),headers={'Content-Type':'application/json','X-SG-Power-Tool':config['token']})
            with opener.open(request,timeout=30) as response:
                raw=response.read(262145)
                if len(raw)>262144:raise ValueError('Fleet power status too large')
                result=json.loads(raw)
            emit('power',event={**event,'state':'complete','finished_at':__import__('datetime').datetime.now(__import__('datetime').timezone.utc).isoformat(),'result':result})
            return json.dumps(result)
        except Exception as error:
            message='Fleet power request could not be confirmed. Read fleet_power_status for the same action ID; never repeat the request or claim it succeeded.'
            if isinstance(error,urllib.error.HTTPError):
                try:message=json.loads(error.read(4096)).get('error',message)
                except (ValueError,OSError):pass
                finally:error.close()
            emit('power',event={**event,'state':'failed','finished_at':__import__('datetime').datetime.now(__import__('datetime').timezone.utc).isoformat(),'error':message})
            return json.dumps({'error':message,'next_step':'Read fleet_power_status and find this action ID; do not issue another start/stop for the same worker.'})
    for name,description,parameters in [
        ('fleet_recipe_rollout','Permanently deploy an exact operator-enrolled Spark recipe after explicit owner authorization. This is not a benchmark or a temporary trial: it copies the previously qualified image, preserves the target serving limits and precision, backs up the original containers/launch configuration, holds only the target pair, waits for existing work, starts the candidate, verifies readiness, updates the persistent launcher and conditionally readmits. The original remains available for rollback. Preparation and deployment run independently after acceptance; read fleet_power_status recipe_trials for the same rollout ID. Never reissue uncertainty with a new ID or claim completion from narration. Only after a confirmed failed copying_qualified_image stage, after addressing its cause, expected_finished_at may resume that same operation. Supply its exact finished_at receipt. Running, uncertain or advanced preparations cannot resume; old failure evidence remains retained.',{'type':'object','properties':{'profile':{'type':'string'},'rollout_id':{'type':'string'},'expected_finished_at':{'type':'number'}},'required':['profile','rollout_id'],'additionalProperties':False}),
        ('fleet_recipe_trial','Run a pre-enrolled, immutable owner-approved recipe qualification. Use the exact profile ID provided by the owner/operator. prepare backs up and verifies the original; Spark profiles also build an isolated candidate while serving remains running. run reserves only its enrolled worker, waits for gateway/native work, executes the enrolled protocol (candidate-only correctness/cache/capacity acceptance or an A/B/A2 comparison), restores the original containers or exact local settings and conditionally readmits. The profile fixes the protocol; chat cannot change it. Never adopt a candidate or change the plan. Use the existing explicit approval for this exact temporary profile; otherwise ask first. Use the SAME trial_id for prepare and run. Preparation must be complete before run. Returns promptly; read fleet_power_status recipe_trials for durable status. Never reissue an uncertain stage with a new ID.',{'type':'object','properties':{'profile':{'type':'string'},'stage':{'type':'string','enum':['prepare','run']},'trial_id':{'type':'string'}},'required':['profile','stage','trial_id'],'additionalProperties':False}),
        ('fleet_routing','For an owner-approved lifecycle action, drain or resume one enrolled worker through the existing core controls. Drain stops new routing and lets admitted work finish; never stops a model. Keep a separate healthy LLM. Save the returned operator_action ID; resume requires that exact expected_operator_action so a later owner pause is preserved. If was_drained is true, preserve that preexisting pause unless explicitly authorized otherwise. Use fleet_power_status to inspect routing receipts after uncertainty. Never repeat an action ID with different arguments.',{'type':'object','properties':{'worker':{'type':'string'},'routing_action':{'type':'string','enum':['drain','resume']},'action_id':{'type':'string'},'expected_operator_action':{'type':['string','null']}},'required':['worker','routing_action','action_id'],'additionalProperties':False}),
        ('fleet_power_status','Read enrolled power scripts, current gateway routing per member and recent script receipts. Use before any start or stop. Gateway health does not prove the model process is running; the status script does.',{'type':'object','properties':{},'additionalProperties':False}),
        ('fleet_power','Run one enrolled script: status, start or stop for one exact worker. Use the owner approval already present in this conversation when it covers this exact action; otherwise ask BEFORE stopping or starting. Read-only status needs no approval. Before a stop: use fleet_routing to drain the worker (then wait for load 0 and queued 0), never stop the last healthy LLM. The receipt is a script exit, not readiness or shutdown proof; follow up with fleet_power_status.',{'type':'object','properties':{'worker':{'type':'string'},'power_action':{'type':'string','enum':['status','start','stop']},'action_id':{'type':'string'}},'required':['worker','power_action','action_id'],'additionalProperties':False})]:
        registry.register(name=name,toolset=TOOLSET,schema={'name':name,'description':description,'parameters':parameters},handler=lambda args,_name=name,**kw:run(_name,args),max_result_size_chars=262144)
    return NAMES

File: ds4-gateway/genie-power.mjs
// Conversational and dashboard fleet power tools: read status, and start/stop
// enrolled models through the shared power runner. Per the owner's autonomy
// model, Genie asks first in conversation; the human approves in that same
// conversation, so no separate per-request approval button exists. A deliberate
// dashboard action supplies the owner's intent directly. The scripts remain the
// source of truth; receipts record script exits and REAL endpoint verification.
import {createToolEndpoint} from './genie-tool-endpoint.mjs';
import {powerWorkers,powerScript,machineGroup} from './power-scripts.mjs';

export function fleetPowerEvidence({runner,workers=[],now=Date.now,catalogue=null}){
  const byId=new Map(workers.map(w=>[w.id,w]));
  const members=powerWorkers().map(id=>{
    const w=byId.get(id);
    return {worker_id:id,machine:machineGroup(id),enrolled:true,busy:runner.busy(id),
      script_available:Object.fromEntries(['status','start','stop'].map(a=>[a,!!(runner.script??powerScript)(id,a)])),
      ...(w?{
        routing:{is_healthy:w.is_healthy===true,drained:w.drained===true,load:w.load??0,queued:w.queued??0,
          direct_reserved:w.direct_reserved===true,quarantined:!!w.quarantine},
        gateway_view:'Gateway routing view only. It does not prove the model process is running; use the status script for that.',
      }:{
        routing:null,
        gateway_view:'Not a current gateway worker; only script status applies.',
      })};
  });
  return {schema:1,observed_at:new Date(now()).toISOString(),members,
    ...(catalogue?{catalogue}:{}),
    recent:runner.receipts().slice(0,8),
    scope:'Enrolled power scripts with physical-machine groups and the last receipts of this dashboard process. Scripts remain the source of truth; start/stop receipts include real endpoint verification, and timeout means unproven, not failed. Stopping a Spark pair stops both machines of that pair, including any other model serving there.'};
}

export function createFleetPowerTools({runner,read,isTesting=()=>false,isEnabled=()=>true,directRunning=null,catalogue=null,control=null,recipes=null}={}){
  if(!runner||typeof read!=='function')throw new Error('Fleet power tools need a script runner and gateway status reader.');
  if(directRunning!==null&&typeof directRunning!=='function')throw new Error('directRunning must be a function when provided');
  if(catalogue!==null&&typeof catalogue!=='function')throw new Error('catalogue must be a function when provided');
  const requests=new Map(),routingReceipts=new Map();
  let admission=Promise.resolve();
  async function snapshotWorkers(){
    const value=await read();
    if(value?.version!==1||!Array.isArray(value.workers))throw new Error('Gateway worker registry is unavailable.');
    return value.workers;
  }
  // Synchronous-in-cost preflight: every refusal here is known before any
  // mutation starts. Returns {allowed,refusals} and never runs scripts.
  async function precheck(worker,power_action){
    const workers=await snapshotWorkers();
    const current=workers.find(w=>w.id===worker);
    if(!current)throw new Error('Worker is not a current gateway member; use the script directly.');
    const refusals=[];
    if(recipes?.busy(worker))refusals.push('An enrolled recipe trial owns this hardware; inspect its restoration receipt before another lifecycle change.');
    if(power_action==='stop'){
      const groups=machineGroup(worker)??[];
      if(current.is_healthy&&!current.drained)refusals.push('Drain the worker before stopping so new gateway work cannot be admitted.');
      if((current.load??0)>0||current.queued>0)refusals.push('Worker is serving or holds queued gateway work. Drain it first (drain-workers), let admitted work finish, then stop.');
      if(current.direct_reserved===true)refusals.push('The endpoint is in active direct owner use right now; stopping would interrupt it.');
      if(typeof directRunning==='function'){
        const running=await directRunning(worker);
        if(!Number.isFinite(running)&&current.is_healthy)refusals.push('Native activity is unknown; obtain a fresh idle observation before stopping.');
        if(Number.isFinite(running)&&running>0)refusals.push(`The engine reports ${running} running request(s) outside gateway accounting; treat them as active work, not idle.`);
      }
      const mates=workers.filter(w=>w.id!==worker&&(machineGroup(w.id)??[]).some(group=>groups.includes(group)));
      const busyMates=mates.filter(w=>(w.load??0)>0||w.queued>0||w.direct_reserved===true||(w.is_healthy&&!w.drained));
      if(busyMates.length)refusals.push(`Same-hardware model(s) ${busyMates.map(w=>w.id).join(', ')} still hold gateway work; stopping this model stops the shared machine(s).`);
      for(const mate of mates.filter(w=>w.is_healthy))if(directRunning){
        const n=await directRunning(mate.id);
        if(!Number.isFinite(n)||n>0)refusals.push(`Same-hardware model ${mate.id} has active or unknown native work.`);
      }
      const healthy=workers.filter(w=>w.is_healthy&&!w.drained&&!runner.busy(w.id)&&!(machineGroup(w.id)??[]).some(group=>groups.includes(group)));
      if(!healthy.length)refusals.push('This hardware contains the last healthy worker. Stopping it leaves no LLM; arrange a replacement first.');
    }
    return {allowed:refusals.length===0,refusals,worker,power_action,
      machine:machineGroup(worker),
      scope:power_action==='stop'?'Stopping a Spark pair stops both machines of that pair, including any other model serving there.':'Starting may conflict with a different model already serving the same machine; the script refuses that case and reports it.'};
  }
  async function runTool(input){
    if(input?.action==='status'&&Object.keys(input).length===1){
      const cat=catalogue?await catalogue().catch(e=>({unavailable:e.message})):null;
      return {...fleetPowerEvidence({runner,workers:await snapshotWorkers(),catalogue:cat}),routing_recent:[...routingReceipts.values()].slice(-16).reverse(),recipe_trials:recipes?.status()??[]};
    }
    if(input?.action==='recipe-trial'){
      if(Object.keys(input).sort().join(',')!=='action,profile,stage,trial_id'||!recipes)throw Error('Use an enrolled recipe trial profile and stage');
      if(isTesting()||!isEnabled())throw Error('Recipe trials are suspended or fleet power is switched off');
      return recipes.start(input);
    }
    if(input?.action==='routing'){
      const {worker,routing_action,action_id,expected_operator_action}=input;
      const keys=Object.keys(input).sort().join(',');
      if(!['action,action_id,routing_action,worker','action,action_id,expected_operator_action,routing_action,worker'].includes(keys)||!['drain','resume'].includes(routing_action)||!/^[a-f0-9-]{36}$/.test(action_id??''))throw Error('Specify an exact worker, routing action and action ID.');
      const previous=routingReceipts.get(action_id);
      if(previous){if(previous.worker!==worker||previous.routing_action!==routing_action)throw Error('Action ID belongs to another routing action');return previous;}
      if(requests.has(action_id))throw Error('Action ID belongs to a power action');
      if(isTesting()||!isEnabled()||!control)throw Error('Fleet routing control is unavailable or switched off.');
      const workers=await snapshotWorkers(),current=workers.find(w=>w.id===worker);
      if(!current||!machineGroup(worker))throw Error('Use an enrolled current fleet worker');
      if(routing_action==='drain'&&!workers.some(w=>w.id!==worker&&w.is_healthy&&!w.drained&&!runner.busy(w.id)&&!(machineGroup(w.id)??[]).some(g=>(machineGroup(worker)??[]).includes(g))))throw Error('Keep a healthy worker on separate hardware before draining.');
      if(routing_action==='resume'&&(runner.busy(worker)||!Object.hasOwn(input,'expected_operator_action')||(expected_operator_action!==null&&!/^[a-f0-9-]{36}$/.test(expected_operator_action))))throw Error('Wait for power completion and supply the observed operator-action ID before resuming.');
      const receipt={worker,routing_action,action_id,state:'running',was_drained:current.drained===true};routingReceipts.set(action_id,receipt);
      try{
        await control(routing_action==='drain'?'/drain-workers':'/resume-workers',{workers:[worker],...(routing_action==='resume'?{expected_operator_actions:{[worker]:expected_operator_action}}:{})});
        const after=(await snapshotWorkers()).find(w=>w.id===worker);
        if(!after||after.drained!==(routing_action==='drain'))throw Error('Routing state was not confirmed');
        Object.assign(receipt,{state:'complete',drained:after.drained,load:after.load,queued:after.queued,operator_action:after.last_operator_action?.id??null,scope:'Routing only; existing work continues and the model process is unchanged. Preserve preexisting pauses.'});
      }catch(error){Object.assign(receipt,{state:'unverified',error:error.message});}
      return receipt;
    }
    const keys=Object.keys(input??{}).sort().join(',');
    if(input?.action!=='power'||!['action,action_id,power_action,worker','action,action_id,mode,power_action,worker'].includes(keys)||('mode' in input&&input.mode!=='check'))
      throw new Error('Specify worker, power_action and one action ID.');
    const {worker,power_action,action_id}=input;
    if(!/^[a-f0-9-]{36}$/.test(action_id??''))throw new Error('Provide one action ID for this power request.');
    if(!['start','stop','status'].includes(power_action))throw new Error('power_action must be start, stop or status.');
    if(!(runner.script??powerScript)(worker,power_action))throw new Error(`No enrolled script for ${worker} ${power_action}; scripts remain the source of truth.`);
    if(routingReceipts.has(action_id))throw Error('Action ID belongs to a routing action');
    const prior=requests.get(action_id);
    if(prior){
      if(prior.worker!==worker||prior.power_action!==power_action)throw new Error('Action ID already belongs to a different request.');
      return prior.result??{accepted:true,action_id,worker,power_action,state:'running',next_step:'Read fleet_power_status for this action ID. Do not repeat the mutation.'};
    }
    if(power_action!=='status'){
      if(isTesting())throw new Error('Fleet power is suspended for testing.');
      if(!isEnabled())throw new Error('Fleet power is switched off.');
    }
    if(power_action==='status'){
      if(input.mode==='check')return {allowed:true,action_id,worker,power_action,machine:machineGroup(worker),scope:'Read-only status script.'};
      const receipt=await runner.run(worker,'status',{actionId:action_id});
      return {receipt,action_id,next_step:'Script output is in the receipt; it describes serving engine and containers, not gateway routing.'};
    }
    const pre=await precheck(worker,power_action);
    if(!pre.allowed)throw new Error(pre.refusals.join(' '));
    if(input.mode==='check')return {allowed:true,action_id,worker,power_action,machine:pre.machine,scope:pre.scope,
      note:'Preflight only. The mutation runs through the same runner as Genie and reports real endpoint verification.'};
    if(requests.has(action_id))return runTool(input);
    const request={worker,power_action};requests.set(action_id,request);
    void runner.run(worker,power_action,{actionId:action_id}).then(receipt=>{request.result={receipt,action_id,
      next_step:power_action==='stop'
        ?receipt.verified?.state==='stopped'
          ?'Verified stopped: the endpoint no longer accepts connections. Gateway routing will mark it unavailable on its next probes.'
          :'Stop receipt recorded but shutdown is NOT verified. Read fleet_power_status or the card Status; do not claim the model is down.'
        :receipt.verified?.state==='ready'
          ?'Verified ready: the endpoint answers model-list requests. Routing needs the gateway probes to mark it healthy.'
          :'Start receipt recorded but readiness is NOT verified (possibly still loading). Poll Status; do not route expectations yet.'};
    }).catch(error=>{request.result={action_id,worker,power_action,error:error.message,state:'unverified'};});
    return {accepted:true,action_id,worker,power_action,state:'running',next_step:'Read fleet_power_status for this action ID. Accepted means the operation is running, not verified complete.'};
  }
  const tool=input=>{
    if(input?.action==='recipe-trial'||input?.action==='routing'||(input?.action==='power'&&input.power_action!=='status'&&input.mode!=='check')){
      const next=admission.then(()=>runTool(input));admission=next.then(()=>undefined,()=>undefined);return next;
    }
    return runTool(input);
  };
  const endpoint=createToolEndpoint('/api/genie/power-tools','x-sg-power-tool',tool);
  return {...endpoint,precheck:(worker,power_action)=>precheck(worker,power_action),tool};
}
